File: day14/main.go
package main

import (
	"bytes"
	"fmt"
)

const (
	puzzleInput  = 540391
	puzzleInput2 = "540391"
)

func split(b byte) (byte, byte) {
	return b / 10, b % 10
}

func cmp(state, target []byte) int {
	sl := len(state)
	tl := len(target)
	if sl < tl+1 {
		return -1
	}
	l := sl - tl
	if bytes.Equal(state[l-1:sl-1], target) || bytes.Equal(state[l:], target) {
		return l
	}
	return -1
}

func main() {
	count1 := 0
	count2 := 1
	state := []byte{3, 7}
	for len(state) < puzzleInput+10 {
		a, b := split(state[count1] + state[count2])
		if a > 0 {
			state = append(state, a)
		}
		state = append(state, b)
		count1 = (count1 + int(state[count1]) + 1) % len(state)
		count2 = (count2 + int(state[count2]) + 1) % len(state)
	}
	out := make([]byte, 10)
	copy(out, state[puzzleInput:puzzleInput+10])
	for i := range out {
		out[i] += byte('0')
	}
	fmt.Println(string(out))

	target := []byte(puzzleInput2)
	for n, i := range target {
		target[n] = i - byte('0')
	}
	state = []byte{3, 7}
	count1 = 0
	count2 = 1
	k := -1
	for ; k < 0; k = cmp(state, target) {
		a, b := split(state[count1] + state[count2])
		if a > 0 {
			state = append(state, a)
		}
		state = append(state, b)
		count1 = (count1 + int(state[count1]) + 1) % len(state)
		count2 = (count2 + int(state[count2]) + 1) % len(state)
	}
	fmt.Println(k)
}

File: day14/src/main.rs
const PUZZLEINPUT: usize = 540391;
const PUZZLEINPUT2: &[u8] = &[5, 4, 0, 3, 9, 1];

fn split(b: u8) -> (u8, u8) {
    (b / 10, b % 10)
}

fn compare(state: &[u8], target: &[u8]) -> usize {
    let sl = state.len();
    let tl = target.len();
    if sl < tl + 1 {
        return 0;
    }
    let l = sl - tl;
    if &state[l - 1..sl - 1] == target || &state[l..] == target {
        return l;
    }
    return 0;
}

fn main() {
    part1();
    part2();
}

fn part1() {
    let mut count1 = 0;
    let mut count2 = 1;
    let mut state = vec![3, 7];
    while state.len() < PUZZLEINPUT + 10 {
        let (a, b) = split(state[count1] + state[count2]);
        if a > 0 {
            state.push(a);
        }
        state.push(b);
        count1 = (count1 + state[count1] as usize + 1) % state.len();
        count2 = (count2 + state[count2] as usize + 1) % state.len();
    }
    println!(
        "{}",
        state[PUZZLEINPUT..PUZZLEINPUT + 10]
            .iter()
            .map(|&x| (x + '0' as u8) as char)
            .collect::<String>()
    );
}

fn part2() {
    let mut count1 = 0;
    let mut count2 = 1;
    let mut state = vec![3, 7];

    let mut k = 0;
    while k == 0 {
        let (a, b) = split(state[count1] + state[count2]);
        if a > 0 {
            state.push(a);
        }
        state.push(b);
        count1 = (count1 + state[count1] as usize + 1) % state.len();
        count2 = (count2 + state[count2] as usize + 1) % state.len();
        k = compare(&state, PUZZLEINPUT2);
    }
    println!("{}", k);
}
